google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceConfigRpcCache.java

```diff
@@ -0,0 +1,121 @@
+                                  if (instance.getEndpointUrisCount() > 0) {
+                                    optionsBuilder.setHost(instance.getEndpointUris(0));
+                                  } else {
+                                    return projectClient;
+                                  }
+                                } catch (SpannerException e) {
+                                  if (e.getErrorCode() == ErrorCode.UNIMPLEMENTED) {
+                                    // Ignore...
+                                    // This is backwards compatibility.
+                                    return projectClient;
+                                  } else {
+                                    logger.log(
+                                            Level.WARNING,
+                                            "Failed while resolving endpoint URLs for instance:"
+                                                    + instanceName.toString()
+                                                    + " reason: "
+                                                    + e.getErrorCode().name());
+                                    throw e;
+                                  }
+                                } catch (Exception e) {
+                                  logger.log(
+                                          Level.WARNING,
+                                          "Failed while resolving endpoint URLs for instance:"
+                                                  + instanceName.toString());
+                                  throw SpannerExceptionFactory.newSpannerException(e);
+                                }
+                                return new GapicSpannerRpc(optionsBuilder.build(), false);
+                              }
+                            });
+  }
+
+  public SpannerRpc get(SessionName sessionName) {
+    InstanceName instanceName =
+            InstanceName.of(sessionName.getProject(), sessionName.getInstance());
+    return get(instanceName);
+  }
+
+  public SpannerRpc get(DatabaseName databaseName) {
+    InstanceName instanceName =
+            InstanceName.of(databaseName.getProject(), databaseName.getInstance());
+    return get(instanceName);
+  }
+
+  private SpannerRpc get(InstanceName instanceName) {
+    try {
+      SpannerRpc spannerRpc = cache.get(instanceName);
+      return spannerRpc;
+    } catch (ExecutionException e) {
+      logger.log(
+              Level.FINE, "Failed looking up instance in cache. id:" + instanceName.toString(), e);
+      throw SpannerExceptionFactory.newSpannerException(
+              ErrorCode.NOT_FOUND,
+              "Failed getting RPC Client for Instance:" + instanceName.toString(),
+              e);
+    }
+  }
+
+  public void invalidateAll() {
+    for (SpannerRpc rpc : cache.asMap().values()) {
+      if (rpc == this.projectClient) continue;
+      rpc.shutdown();
+    }
+    cache.invalidateAll();
+  }
+}
```

Comment:
nit: missing line ending

Remember to run the code formatter like this:
`mvn com.coveo:fmt-maven-plugin:format`